File: simulation/quantum/quantum_walk.py
"""
Quantum Random Walk for Opinion Diffusion (Cancel Culture / Viral Moral Panic).

Classical random walk diffuses as O(√t); quantum walk diffuses as O(t).
Models the rapid spread of moral judgment in social networks (e.g. cancel culture).
"""

from typing import Tuple, Optional, List
import numpy as np


def _hadamard_1d() -> np.ndarray:
    """2x2 Hadamard gate."""
    return np.array([[1, 1], [1, -1]], dtype=complex) / np.sqrt(2)


def quantum_walk_step(
    state: np.ndarray,
    n_sites: int,
    coin_bias: float = 0.5,
) -> np.ndarray:
    """
    Single step of Hadamard quantum walk on a line.

    Parameters
    ----------
    state : ndarray, shape (2 * n_sites,)
        Combined position-coin state (position 0..n_sites-1, coin 0 or 1).
    n_sites : int
        Number of sites (graph nodes).
    coin_bias : float, default=0.5
        Bias toward left (0) vs right (1) movement. 0.5 = unbiased.

    Returns
    -------
    ndarray
        New state after one step.
    """
    H = _hadamard_1d() if abs(coin_bias - 0.5) < 1e-9 else np.array([
        [np.sqrt(coin_bias), np.sqrt(1 - coin_bias)],
        [np.sqrt(1 - coin_bias), -np.sqrt(coin_bias)],
    ], dtype=complex)
    new_state = np.zeros_like(state)
    for p in range(n_sites):
        for c in range(2):
            idx = p * 2 + c
            amp = state[idx]
            if abs(amp) < 1e-15:
                continue
            for cn in range(2):
                dest = p + (1 if cn == 1 else -1)
                if 0 <= dest < n_sites:
                    new_state[dest * 2 + cn] += H[cn, c] * amp
    return new_state


def quantum_walk_propagate(
    n_sites: int = 31,
    steps: int = 10,
    initial_site: Optional[int] = None,
    coin_bias: float = 0.5,
    seed: Optional[int] = None,
) -> Tuple[np.ndarray, np.ndarray]:
    """
    Propagate quantum walk and return position probability distribution.

    Parameters
    ----------
    n_sites : int, default=31
        Number of sites on the line.
    steps : int, default=10
        Number of walk steps.
    initial_site : int | None, default=None
        Initial position. If None, center.
    coin_bias : float, default=0.5
        Coin bias.
    seed : int | None, default=None
        Random seed (unused for unitary evolution).

    Returns
    -------
    Tuple[ndarray, ndarray]
        (position_probs, timesteps) - prob[i,t] = P(position=i at step t)
    """
    if initial_site is None:
        initial_site = n_sites // 2
    state = np.zeros(2 * n_sites, dtype=complex)
    state[initial_site * 2] = 1.0  # |position, coin=0⟩

    probs = np.zeros((n_sites, steps + 1))
    probs[initial_site, 0] = 1.0

    for t in range(1, steps + 1):
        state = quantum_walk_step(state, n_sites, coin_bias)
        for p in range(n_sites):
            probs[p, t] = float(np.abs(state[p * 2]) ** 2 + np.abs(state[p * 2 + 1]) ** 2)

    return probs, np.arange(steps + 1)


def diffusion_spread(
    n_sites: int = 31,
    steps: int = 20,
    classical: bool = False,
    seed: Optional[int] = None,
) -> Tuple[float, float]:
    """
    Compare classical vs quantum diffusion spread (std of distribution).

    Classical: std ~ O(√t)
    Quantum: std ~ O(t)

    Parameters
    ----------
    n_sites : int
    steps : int
    classical : bool
        If True, simulate classical random walk (binomial).
    seed : int | None

    Returns
    -------
    Tuple[float, float]
        (final_std, spread_at_last_step)
    """
    rng = np.random.default_rng(seed)
    center = n_sites // 2
    if classical:
        pos = center
        positions = [pos]
        for _ in range(steps - 1):
            pos += 1 if rng.random() > 0.5 else -1
            pos = max(0, min(n_sites - 1, pos))
            positions.append(pos)
        probs = np.zeros((n_sites, steps))
        for t, p in enumerate(positions):
            probs[p, t] = 1.0
        final_dist = np.sum(probs, axis=1) / steps
    else:
        probs, _ = quantum_walk_propagate(n_sites, steps, initial_site=center)
        final_dist = probs[:, -1]
    x = np.arange(n_sites)
    mean = np.sum(x * final_dist) / (np.sum(final_dist) + 1e-15)
    std = np.sqrt(np.sum((x - mean) ** 2 * final_dist) / (np.sum(final_dist) + 1e-15))
    return float(std), float(np.max(final_dist))


def simulate_cancel_culture_spread(
    n_agents: int = 50,
    steps: int = 15,
    initial_infected: int = 1,
    quantum: bool = True,
    seed: Optional[int] = None,
) -> dict:
    """
    Simulate opinion/moral panic spread (cancel culture analogy).

    Uses quantum walk for O(t) diffusion when quantum=True,
    classical for O(√t) when quantum=False.

    Returns
    -------
    dict
        - spread_at_step: list of std at each step
        - infected_fraction: fraction of agents with non-negligible probability
        - quantum: bool
    """
    if quantum:
        probs, timesteps = quantum_walk_propagate(
            n_sites=n_agents,
            steps=steps,
            initial_site=min(initial_infected, n_agents - 1),
            seed=seed,
        )
        spread_at_step = []
        for t in range(steps + 1):
            dist = probs[:, t]
            x = np.arange(n_agents)
            mean = np.sum(x * dist) / (np.sum(dist) + 1e-15)
            std = np.sqrt(np.sum((x - mean) ** 2 * dist) / (np.sum(dist) + 1e-15))
            spread_at_step.append(float(std))
        infected_fraction = np.sum(probs[:, -1] > 0.01) / n_agents
    else:
        rng = np.random.default_rng(seed)
        positions = [np.zeros(n_agents) for _ in range(initial_infected)]
        for i in range(initial_infected):
            p = min(i, n_agents - 1)
            positions[i][p] = 1.0
        # Simple classical diffusion: each step, each agent passes to neighbors
        state = np.zeros(n_agents)
        for i in range(initial_infected):
            state += positions[i]
        state /= initial_infected
        spread_at_step = []
        for _ in range(steps + 1):
            x = np.arange(n_agents)
            mean = np.sum(x * state) / (np.sum(state) + 1e-15)
            std = np.sqrt(np.sum((x - mean) ** 2 * state) / (np.sum(state) + 1e-15))
            spread_at_step.append(float(std))
            # Diffusion step
            new_state = np.zeros_like(state)
            for i in range(n_agents):
                if i > 0:
                    new_state[i - 1] += 0.5 * state[i]
                if i < n_agents - 1:
                    new_state[i + 1] += 0.5 * state[i]
            state = new_state
        infected_fraction = np.sum(state > 0.01) / n_agents

    return {
        "spread_at_step": spread_at_step,
        "infected_fraction": float(infected_fraction),
        "quantum": quantum,
    }


# ---------------------------------------------------------------------------
# Coined Quantum Walk on an arbitrary (belief) graph  --  §1.2 of plans
# ---------------------------------------------------------------------------

from typing import Dict


class GraphQuantumWalk:
    """
    Szegedy-style coined quantum walk on a graph representing the "belief space"
    of moral agents.  Each vertex is a stance, each edge an allowed transition.

    The walker's state lives on *directed* edges: H_walker = C^{|E_directed|}.
    A coin operator (Grover diffusion across the edges of each vertex) mixes
    amplitudes leaving each vertex; the shift operator swaps the pair
    (u -> v) <-> (v -> u).  After t steps, the probability of being at vertex
    v is the sum of |amp(u -> v)|^2 over all u.

    This gives the quadratic (O(t)) diffusion signature that the plan requires
    while also handling irregular graphs (unlike the 1D line in the existing
    `quantum_walk_propagate`).
    """

    def __init__(self, adjacency: Dict[int, List[int]]) -> None:
        self.adjacency = {int(k): [int(v) for v in vs] for k, vs in adjacency.items()}
        self.vertices = sorted(self.adjacency.keys())
        # Directed-edge index: (src, dst) -> int
        self.edge_index: Dict[Tuple[int, int], int] = {}
        for u in self.vertices:
            for v in self.adjacency[u]:
                self.edge_index[(u, v)] = len(self.edge_index)
        self.n_edges = len(self.edge_index)

    # ------------------------------------------------------------------
    def _coin(self, amps: np.ndarray) -> np.ndarray:
        """Grover coin applied independently at every vertex."""
        out = np.zeros_like(amps)
        for u in self.vertices:
            neighbours = self.adjacency[u]
            if not neighbours:
                continue
            indices = [self.edge_index[(u, v)] for v in neighbours]
            local = amps[indices]
            avg = local.mean()
            # Grover: 2|s><s| - I
            out[indices] = 2.0 * avg - local
        return out

    def _shift(self, amps: np.ndarray) -> np.ndarray:
        """Flip (u -> v) amplitude with (v -> u)."""
        out = np.zeros_like(amps)
        for (u, v), idx in self.edge_index.items():
            if (v, u) in self.edge_index:
                out[self.edge_index[(v, u)]] = amps[idx]
        return out

    def step(self, amps: np.ndarray) -> np.ndarray:
        return self._shift(self._coin(amps))

    def initial_state(
        self,
        start_vertex: int,
        phases: Optional[Sequence[float]] = None,
    ) -> np.ndarray:
        """
        Superposition over outgoing edges of start_vertex.

        By default uses a *chirality-breaking* pattern: first outgoing edge
        weighted by (1+i)/sqrt(2), remaining edges by (1-i)/sqrt(2*(k-1)).
        This breaks the perfect symmetry that would otherwise trap the walker
        and reproduces the ballistic (~t) spread expected from a coined walk.
        Pass ``phases`` explicitly to override.
        """
        state = np.zeros(self.n_edges, dtype=complex)
        out_edges = [self.edge_index[(start_vertex, v)] for v in self.adjacency[start_vertex]]
        if not out_edges:
            raise ValueError(f"vertex {start_vertex} has no outgoing edges")
        k = len(out_edges)
        if phases is None:
            if k == 1:
                state[out_edges[0]] = 1.0
                return state
            # Chirality-breaking initial coin:
            #   weight first edge with amplitude  a = (1 + i) / sqrt(2)
            #   remaining edges with amplitude    b = (1 - i) / sqrt(2 (k-1))
            a = (1 + 1j) / np.sqrt(2.0)
            b = (1 - 1j) / np.sqrt(2.0 * (k - 1))
            state[out_edges[0]] = a / np.sqrt(2)
            for idx in out_edges[1:]:
                state[idx] = b / np.sqrt(2)
            # normalise
            state = state / np.linalg.norm(state)
            return state
        if len(phases) != k:
            raise ValueError("phases length must match outgoing degree")
        for idx, phi in zip(out_edges, phases):
            state[idx] = np.exp(1j * phi) / np.sqrt(k)
        return state

    def probabilities(self, amps: np.ndarray) -> Dict[int, float]:
        """Probability of occupying each vertex = sum |amp(*, v)|^2."""
        probs: Dict[int, float] = {v: 0.0 for v in self.vertices}
        for (u, v), idx in self.edge_index.items():
            probs[v] += float(np.abs(amps[idx]) ** 2)
        return probs

    def run(self, start_vertex: int, steps: int) -> List[Dict[int, float]]:
        state = self.initial_state(start_vertex)
        history = [self.probabilities(state)]
        for _ in range(steps):
            state = self.step(state)
            history.append(self.probabilities(state))
        return history


def classical_walk_on_graph(
    adjacency: Dict[int, List[int]],
    start_vertex: int,
    steps: int,
    seed: Optional[int] = None,
) -> List[Dict[int, float]]:
    """Reference classical random walk on the same graph (uniform transitions)."""
    vertices = sorted(adjacency.keys())
    probs = {v: 0.0 for v in vertices}
    probs[start_vertex] = 1.0
    history = [probs.copy()]
    for _ in range(steps):
        new = {v: 0.0 for v in vertices}
        for v in vertices:
            if probs[v] == 0 or not adjacency[v]:
                if not adjacency[v]:
                    new[v] += probs[v]  # absorb
                continue
            share = probs[v] / len(adjacency[v])
            for u in adjacency[v]:
                new[u] += share
        probs = new
        history.append(probs.copy())
    return history


def variance_on_cycle(
    n: int,
    steps: int,
    quantum: bool = True,
) -> float:
    """
    Second moment E[d^2] of the shortest-arc distance from the start vertex
    after `steps` steps on a cycle graph with n vertices.

    Because a coined quantum walk on a cycle spreads symmetrically, the mean
    signed displacement is zero and the classical-vs-quantum separation shows
    up in the *second moment* of the distance.  Classical walks give
    E[d^2] ~ t, quantum walks give E[d^2] ~ t^2 (ballistic spread).
    """
    adjacency = {i: [(i - 1) % n, (i + 1) % n] for i in range(n)}
    start = n // 2
    if quantum:
        history = GraphQuantumWalk(adjacency).run(start, steps)
    else:
        history = classical_walk_on_graph(adjacency, start, steps)
    probs = history[-1]
    total = sum(probs.values())
    if total <= 0:
        return 0.0
    m2 = 0.0
    for v, p in probs.items():
        dist = min(abs(v - start), n - abs(v - start))
        m2 += p * dist * dist / total
    return float(m2)

File: julia/src/QuantumWalk.jl
"""
    QuantumWalk — Quantum Random Walk for Opinion Diffusion (Cancel Culture Model).

    Mirrors simulation/quantum/quantum_walk.py using matrix-based coin+shift operators.

    Classical walk: std ~ O(√t)
    Quantum walk:   std ~ O(t)   ← models rapid moral-panic spread

    Design (coin + shift operator pattern):
      State space: position ∈ {0..n_sites-1} × coin ∈ {0,1}
      State vector: length 2*n_sites, interleaved as [|p,0⟩, |p,1⟩] for each p.
      Coin operator C: Hadamard (or biased) acting on coin degree of freedom.
      Shift operator S: |p,0⟩ → |p-1,0⟩ (left), |p,1⟩ → |p+1,1⟩ (right).
      One step = S · (I_position ⊗ C).
"""
module QuantumWalk

using LinearAlgebra
using Statistics
using Random

export quantum_walk_step, quantum_walk_propagate, diffusion_spread
export simulate_cancel_culture_spread
export build_coin_operator, build_shift_operator

# ---------------------------------------------------------------------------
# Coin operators
# ---------------------------------------------------------------------------

"""
    hadamard_coin() -> Matrix{ComplexF64}

Standard 2×2 Hadamard coin: H = (1/√2) [[1,1],[1,-1]].
"""
function hadamard_coin()::Matrix{ComplexF64}
    return ComplexF64[1 1; 1 -1] ./ sqrt(2.0)
end

"""
    biased_coin(coin_bias) -> Matrix{ComplexF64}

Biased coin matrix with probability `coin_bias` toward left (|0⟩) movement.
C = [[√p, √(1-p)], [√(1-p), -√p]]
"""
function biased_coin(coin_bias::Float64)::Matrix{ComplexF64}
    p  = coin_bias
    q  = 1.0 - coin_bias
    return ComplexF64[sqrt(p) sqrt(q); sqrt(q) -sqrt(p)]
end

"""
    build_coin_operator(n_sites, coin_bias) -> Matrix{ComplexF64}

Full coin operator on 2*n_sites-dimensional state space.
Applies the 2×2 coin to the coin register at each position site independently.
"""
function build_coin_operator(n_sites::Int, coin_bias::Float64=0.5)::Matrix{ComplexF64}
    coin = abs(coin_bias - 0.5) < 1e-9 ? hadamard_coin() : biased_coin(coin_bias)
    d = 2 * n_sites
    C = zeros(ComplexF64, d, d)
    for p in 0:(n_sites-1)
        base = p * 2 + 1  # Julia 1-indexed
        C[base:base+1, base:base+1] .= coin
    end
    return C
end

"""
    build_shift_operator(n_sites) -> Matrix{ComplexF64}

Conditional shift operator on 2*n_sites-dimensional state space.
|p, coin=0⟩ → |p-1, 0⟩  (left shift, reflecting at boundaries)
|p, coin=1⟩ → |p+1, 1⟩  (right shift, reflecting at boundaries)
"""
function build_shift_operator(n_sites::Int)::Matrix{ComplexF64}
    d = 2 * n_sites
    S = zeros(ComplexF64, d, d)
    for p in 0:(n_sites-1)
        # coin=0 → shift left
        dest_left = p - 1
        if 0 <= dest_left < n_sites
            S[dest_left * 2 + 1, p * 2 + 1] = 1.0  # |dest,0⟩ ← |p,0⟩
        end
        # coin=1 → shift right
        dest_right = p + 1
        if 0 <= dest_right < n_sites
            S[dest_right * 2 + 2, p * 2 + 2] = 1.0  # |dest,1⟩ ← |p,1⟩
        end
    end
    return S
end

# ---------------------------------------------------------------------------
# Core walk step
# ---------------------------------------------------------------------------

"""
    quantum_walk_step(state, n_sites; coin_bias) -> Vector{ComplexF64}

Single step of Hadamard quantum walk on a line.

State layout: state[p*2+1] = amplitude for |p, coin=0⟩ (1-indexed).
              state[p*2+2] = amplitude for |p, coin=1⟩.

Uses coin operator then shift operator: ψ' = S · C · ψ.
"""
function quantum_walk_step(state::Vector{<:Complex}, n_sites::Int;
                            coin_bias::Float64=0.5)::Vector{ComplexF64}
    coin = abs(coin_bias - 0.5) < 1e-9 ? hadamard_coin() : biased_coin(coin_bias)
    new_state = zeros(ComplexF64, 2 * n_sites)

    for p in 0:(n_sites-1)
        amp0 = state[p * 2 + 1]
        amp1 = state[p * 2 + 2]
        abs(amp0) < 1e-15 && abs(amp1) < 1e-15 && continue

        # Apply coin to (amp0, amp1)
        coined = coin * [amp0; amp1]

        # Shift: coined[1] goes left (coin=0), coined[2] goes right (coin=1)
        dest_left  = p - 1
        dest_right = p + 1

        if 0 <= dest_left < n_sites
            new_state[dest_left * 2 + 1] += coined[1]
        end
        if 0 <= dest_right < n_sites
            new_state[dest_right * 2 + 2] += coined[2]
        end
    end
    return new_state
end

# ---------------------------------------------------------------------------
# Propagation
# ---------------------------------------------------------------------------

"""
    quantum_walk_propagate(; n_sites, steps, initial_site, coin_bias, seed)
        -> (probs, timesteps)

Propagate quantum walk and return position probability distribution.

Returns:
  probs[p, t+1]  = P(position=p at step t)   (1-indexed, shape n_sites × steps+1)
  timesteps      = 0:steps
"""
function quantum_walk_propagate(;
    n_sites::Int=31,
    steps::Int=10,
    initial_site::Union{Int,Nothing}=nothing,
    coin_bias::Float64=0.5,
    seed::Union{Int,Nothing}=nothing,
)::Tuple{Matrix{Float64}, Vector{Int}}

    init_pos = initial_site === nothing ? n_sites ÷ 2 : initial_site
    init_pos = clamp(init_pos, 0, n_sites - 1)

    state = zeros(ComplexF64, 2 * n_sites)
    state[init_pos * 2 + 1] = 1.0 + 0im  # |init_pos, coin=0⟩

    probs = zeros(Float64, n_sites, steps + 1)
    probs[init_pos + 1, 1] = 1.0  # t=0 (1-indexed position)

    for t in 1:steps
        state = quantum_walk_step(state, n_sites; coin_bias=coin_bias)
        for p in 0:(n_sites-1)
            probs[p + 1, t + 1] = abs2(state[p * 2 + 1]) + abs2(state[p * 2 + 2])
        end
    end

    return probs, collect(0:steps)
end

# ---------------------------------------------------------------------------
# Diffusion spread comparison
# ---------------------------------------------------------------------------

"""
    diffusion_spread(; n_sites, steps, classical, seed) -> (final_std, max_prob)

Compare classical vs quantum diffusion spread (std of distribution).

Classical: std ~ O(√t)
Quantum:   std ~ O(t)
"""
function diffusion_spread(;
    n_sites::Int=31,
    steps::Int=20,
    classical::Bool=false,
    seed::Union{Int,Nothing}=nothing,
)::Tuple{Float64,Float64}

    rng = seed === nothing ? Random.default_rng() : MersenneTwister(seed)
    center = n_sites ÷ 2

    if classical
        pos = center
        counts = zeros(Float64, n_sites)
        for _ in 1:steps
            pos += rand(rng) > 0.5 ? 1 : -1
            pos = clamp(pos, 0, n_sites - 1)
        end
        final_dist = zeros(Float64, n_sites)
        final_dist[pos + 1] = 1.0
    else
        probs, _ = quantum_walk_propagate(
            n_sites=n_sites, steps=steps, initial_site=center,
            coin_bias=0.5, seed=seed,
        )
        final_dist = probs[:, end]
    end

    s = sum(final_dist)
    s < 1e-15 && return (0.0, 0.0)
    final_dist ./= s

    x = collect(0.0:(n_sites-1))
    mean_pos = sum(x .* final_dist)
    variance = sum((x .- mean_pos) .^ 2 .* final_dist)
    std_dev = sqrt(variance)

    return (std_dev, maximum(final_dist))
end

# ---------------------------------------------------------------------------
# Cancel culture spread simulation
# ---------------------------------------------------------------------------

"""
    simulate_cancel_culture_spread(; n_agents, steps, initial_infected, quantum, seed)
        -> Dict

Simulate opinion / moral-panic spread using quantum or classical random walk.

Quantum walk: O(t) diffusion — models rapid viral spread.
Classical:    O(√t) diffusion — slower baseline.

Returns:
  - spread_at_step: Vector{Float64} — std of position distribution at each step
  - infected_fraction: Float64 — fraction of agents with prob > 0.01 at final step
  - quantum: Bool
"""
function simulate_cancel_culture_spread(;
    n_agents::Int=50,
    steps::Int=15,
    initial_infected::Int=1,
    quantum::Bool=true,
    seed::Union{Int,Nothing}=nothing,
)::Dict{String,Any}

    rng = seed === nothing ? Random.default_rng() : MersenneTwister(seed)

    spread_at_step = Float64[]
    infected_fraction = 0.0

    if quantum
        init_site = clamp(initial_infected, 0, n_agents - 1)
        probs, _ = quantum_walk_propagate(
            n_sites=n_agents, steps=steps, initial_site=init_site,
            coin_bias=0.5, seed=seed,
        )
        for t in 0:steps
            dist = probs[:, t + 1]
            s = sum(dist)
            s < 1e-15 && (push!(spread_at_step, 0.0); continue)
            dist_norm = dist ./ s
            x = collect(0.0:(n_agents-1))
            mean_pos = sum(x .* dist_norm)
            std_dev = sqrt(sum((x .- mean_pos) .^ 2 .* dist_norm))
            push!(spread_at_step, std_dev)
        end
        infected_fraction = sum(probs[:, end] .> 0.01) / n_agents

    else
        # Classical random walk diffusion
        state = zeros(Float64, n_agents)
        init_pos = clamp(initial_infected - 1, 0, n_agents - 1)
        state[init_pos + 1] = 1.0

        x = collect(0.0:(n_agents-1))
        for _ in 0:steps
            s = sum(state)
            if s < 1e-15
                push!(spread_at_step, 0.0)
            else
                dist_norm = state ./ s
                mean_pos = sum(x .* dist_norm)
                std_dev  = sqrt(sum((x .- mean_pos) .^ 2 .* dist_norm))
                push!(spread_at_step, std_dev)
            end
            # Diffuse step: each agent passes 50% to neighbors
            new_state = zeros(Float64, n_agents)
            for i in 1:n_agents
                i > 1       && (new_state[i-1] += 0.5 * state[i])
                i < n_agents && (new_state[i+1] += 0.5 * state[i])
            end
            state = new_state
        end
        infected_fraction = sum(state .> 0.01) / n_agents
    end

    return Dict{String,Any}(
        "spread_at_step"    => spread_at_step,
        "infected_fraction" => infected_fraction,
        "quantum"           => quantum,
    )
end

end # module QuantumWalk
